wireguard: Add fallback to userspace implementation

gandro opened the pull request:
WireGuard on Linux usually uses the kernel-mode implementation, as it is
faster and does not rely on a userspace process running to process
packets.

However, some kernels do not ship with WireGuard support, either because
they are too old, or because WireGuard support has explicitly compiled
out. For such cases, we can fall back to the WireGuard user-space
implementation based on a TUN device. While this is not recommended in
production, it can be useful for testing or in cases where some machines
cannot upgrade to a more recent kernel.

The userspace implementation requires a daemon which encrypts and
decrypts the packets sent over the TUN device. In Cilium, this task is
performed by the cilium-agent process which calls into the wireguard-go
library. This has the consequence that if cilium-agent is down,
connectivity between WireGuard peers, and therefore connectivity between
Cilium-managed pods will also be unavailable. Thus, running WireGuard in
userspace mode is not recommeneded for production workloads. 

The fallback to userspace mode must be explicitly enabled via the
`enable-wireguard-userspace-fallback` option.

Review per commit. The first few commits contain drive-by cleanups.

gandro committed:
wireguard: Fix outdated doc comment

Init used to be called after the WireGuard tunnel IP was allocatd,
however this is not the case anymore since commit 0ecab37c. Init is
still called after the watcher initialization, but it does not require
an IP address anymore.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Do not mask error when disabling rp_filter

It should be propagated to the caller, as we exit the function when
rp_filter cannot be disabled.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Add more descriptive error messages

To aid in troubleshooting when Init() fails.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
operator: Remove unused --enable-wireguard flag

The operator does not have any functionality that conditionally depends
on WireGuard anymore.

Fixes: 0ecab37c49a3 ("wireguard: Remove operator")

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
vendor: Add wireguard userspace packages

This commit adds the packages used for the WireGuard userspace
implementation added in a subsequent commit.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Add fallback to userspace implementation

WireGuard on Linux usually uses the kernel-mode implementation, as it is
faster and does not rely on a userspace process running to process
packets.

However, some kernels do not ship with WireGuard support, either because
they are too old, or because WireGuard support has explicitly compiled
out. For such cases, we can fall back to the WireGuard user-space
implementation based on a TUN device. While this is not recommended in
production, it can be useful for testing or in cases where some machines
cannot upgrade to a more recent kernel.

The userspace implementation requires a daemon which encrypts and
decrypts the packets sent over the TUN device. In Cilium, this task is
performed by the cilium-agent process which calls into the wireguard-go
library. This has the consequence that if cilium-agent is down,
connectivity between WireGuard peers, and therefore connectivity between
Cilium-managed pods will also be unavailable. Thus, running WireGuard in
userspace mode is not recommeneded for production workloads.

The fallback to userspace mode must be explicitly enabled via the
`enable-wireguard-userspace-fallback` option.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
doc: Add docs for WireGuard user-space mode

Mentioning the caveats that apply.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

rolinh reviewed (approved):
LGTM :rocket:

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -164,12 +252,12 @@ func (a *Agent) Init(mtuConfig mtu.Configuration) error {
 		rt.Prefix = subnet
 		if _, err := route.Upsert(rt); err != nil {
-			return err
+			return fmt.Errorf("failed to upsert ipv4 rule: %w", err)
```
Nit:
```suggestion
			return fmt.Errorf("failed to upsert ipv4 route: %w", err)
```

gandro replied:
Addressed!

tklauser commented on pkg/option/config.go:
```diff
@@ -1400,6 +1403,9 @@ type DaemonConfig struct {
 	// EnableWireguard enables Wireguard encryption
 	EnableWireguard bool
 
+	// WireguardRequireKernelMode enables the fallback to the userspace implementation
```
Nit: this says "enables the fallback..." while the usage message in `daemon/cmd/daemon_main.go` says "disables the fallback...". I guess this should says "disables" as well:

```suggestion
	// WireguardRequireKernelMode disables the fallback to the userspace implementation
```

gandro replied:
Yes! Good catch!

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -88,30 +93,113 @@ func NewAgent(privKeyPath string) (*Agent, error) {
+	}
+	dev := device.NewDevice(tundev, conn.NewDefaultBind(), logger)
+
+	// cleanup removes the tun device and uapi socket
+	a.cleanup = append(a.cleanup, func() {
+		uapiServer.Close()
+		dev.Close()
+	})
```
Do we need to perform this under `a.Lock()?` given that `(*Agent).Close` takes the lock to run the `cleanup` funcs?

gandro replied:
So upon closer inspection it turns out that Init() can be called concurrently also from the node manager, so it needs to take a.Lock(). So I added an additional bugfix commit that I will likely separately backport as well if it passes review here: https://github.com/cilium/cilium/pull/17451/commits/1a201bf8e052126d3d8ccd06b9e3b64af94a465c

tklauser replied:
> So I added an additional bugfix commit that I will likely separately backport as well if it passes review here: [1a201bf](https://github.com/cilium/cilium/commit/1a201bf8e052126d3d8ccd06b9e3b64af94a465c)

Looks like that commit is not part of this PR's branch. Is this intended? In any case, that commit LGTM.

gandro replied:
Weird. It should part of the branch, at least I see it in GitHub's UI. Note however that it is _not_ the last commit, I rewrote the history to include it earlier.

tklauser replied:
Sorry for the noise. I see the commit now. Must have overlooked it previously and probably was confused by the message in the GitHub UI saying that commit isn't part of any branch.

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -88,30 +93,113 @@ func NewAgent(privKeyPath string) (*Agent, error) {
 
 // Close is called when the agent stops
 func (a *Agent) Close() error {
+	a.Lock()
```
Could this be `a.RLock` given `a.cleanup` is not modified?

gandro replied:
Addressed!

tklauser reviewed (commented):
Not directly related to this PR but something that came to my mind while reviewing:

In case the wireguard kernel implementation is built as a module (which seems to be the default configuration and is the case on the few systems I checked): do we check whether the module is loaded before attempting to set up wireguard or would it make sense to add such a check (similar to the way we check for iptables modules being loaded)? Or is it guaranteed to be autoloaded once we create the `wgX` interface?

gandro committed:
wireguard: Add mutex in Init function

Previously, the Init() function did not lock the WireGuard agent under
the assumption that there can be no concurrent calls into the agent.
That assumption however is false, both Close() (via signal handler
cleanup) and UpdatePeer() (via node manager) can be called concurrently
while Init() is being invoked.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

tklauser reviewed (approved):
:rocket:

qmonnet commented on pkg/wireguard/agent/agent.go:
```diff
@@ -96,7 +96,7 @@ func (a *Agent) Close() error {
 }
 
-// Init is called after we have obtained a local Wireguard IP
+// Init creates and configures the local WireGuard tunnel device.
```
Nit: Commit log for this change says:
> however this is not the case anymore since 508969.

I'm not sure what this number refers to, it doesn't seem to point to a PR or a commit.

gandro replied:
Ah, good catch. It's referring to https://github.com/cilium/cilium/pull/15565/commits/508969a8d0e2d0ab982add064851df63c3a6a8dd but the commit sha is different on master of course. Will fix.

qmonnet reviewed (approved):
Looks good to me, nice work! (And the PR is easy to review :pray:.)

borkmann commented on pkg/wireguard/agent/agent.go:
```diff
@@ -88,14 +93,76 @@ func NewAgent(privKeyPath string) (*Agent, error) {
+	log.WithField(logfields.Hint,
+		"It is highly recommended to use the kernel implementation. "+
+			"See https://www.wireguard.com/install/ for details.").
+		Info("falling back to the WireGuard userspace implementation.")
```
nit: Should this be a Warn instead?

borkmann replied:
I wonder if the fallback should be disabled by default. So that when the agent starts up, it hints to the user that a fallback could be used but with mentioned constraints to be aware of?

gandro replied:
Yes, I considered that too, and I am now also starting to think that we probably should make it opt-in instead of opt-out. It will make the installation slightly more annoying, but it will stop people from running user-mode by accident, which I think can be a footgun due to it's performance and availability downsides.

borkmann replied:
If I understood you correctly, it's that wg would be running as part of the Cilium agent process, so upgrading may shortly interrupt/break connectivity. I think an implicit fallback would not make people aware of the constraints.

gandro replied:
That's correct, upgrading or configuration changes (or anything that restarts the cilium-agent process) will interrupt connectivity in user-mode.

gandro committed:
helm: Add `encryption.wireguard.userspaceFallback` value

To enable the fallback to user-space mode implemented in the previous
commits.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

brb reviewed (approved):
Cool! :shipit:
